QE: Split off cobbler steps in cobbler_steps.rb

ByteOtter opened the pull request:
## What does this PR change?

Split off the large block of Cobbler steps from `common_steps.rb` to `cobbler_steps.rb`.

|From File|Functionality|New File|
|----------------------------|----------------------------|--------------------------|
|`common_steps.rb`|everything Cobbler|`cobbler_steps.rb`|
|`command_steps.rb`|cobbler daemon commands|`cobbler_steps.rb`|

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
- No tests: Only internal restructuring

- [x] **DONE**

## Links

Related to [#20191](https://github.com/SUSE/spacewalk/issues/20191)
Manager-4.3: https://github.com/SUSE/spacewalk/pull/20450
Manager-4.2: https://github.com/SUSE/spacewalk/pull/20471

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)


## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

nodeg reviewed (approved):
LGTM. Nevertheless I would recommend doing a PR test.

elariekerboull commented on testsuite/features/step_definitions/cobbler_steps.rb:
```diff
@@ -0,0 +1,121 @@
+  raise "error in adding parameter and value to Cobbler #{item}.\nLogs:\n#{result}" if code.nonzero?
+end
+
+When(/^I check the Cobbler parameter "([^"]*)" with value "([^"]*)" in the isolinux.cfg$/) do |param, value|
```
Same, it looks to be still present in the other file

ByteOtter replied:
fixing now

elariekerboull commented on testsuite/features/step_definitions/common_steps.rb:
```diff
@@ -754,80 +754,6 @@
-  raise "error in verifying Cobbler buildiso image with xorriso.\nLogs:\n#{out}" if code.nonzero?
-end
-
 When(/^I cleanup xorriso temp files$/) do
```
Same

nodeg reviewed (changes requested):
Nice catch, Erwan! I did count the number of methods and they do not match! Please adjust that @ByteOtter.

nodeg commented:
There is more code that should be moved to the new Cobbler file like e.g.

```ruby
Given(/^distro "([^"]*)" exists$/) do |distro|
  ct = CobblerTest.new
  raise 'distro ' + distro + ' does not exist' unless ct.distro_exists(distro)
end

Given(/^profile "([^"]*)" exists$/) do |profile|
  ct = CobblerTest.new
  raise 'profile ' + profile + ' does not exist' unless ct.profile_exists(profile)
end

When(/^I remove kickstart profiles and distros$/) do
  host = $server.full_hostname
  # -------------------------------
  # Cleanup kickstart distros and their profiles, if any.

  # Get all distributions: created from UI or from API.
  distros = $server.run('cobbler distro list')[0].split

  # The name of distros created in the UI has the form: distro_label + suffix
  user_details = $api_test.user.get_details('testing')
  suffix = ":#{user_details['org_id']}:#{user_details['org_name'].delete(' ')}"

  distros_ui = distros.select { |distro| distro.end_with? suffix }.map { |distro| distro.split(':')[0] }
  distros_api = distros.reject { |distro| distro.end_with? suffix }
  distros_ui.each { |distro| $api_test.kickstart.tree.delete_tree_and_profiles(distro) }
  # -------------------------------
  # Remove profiles and distros created with the API.

  # We have already deleted the profiles from the UI; delete all the remaning ones.
  profiles = $server.run('cobbler profile list')[0].split
  profiles.each { |profile| $server.run("cobbler profile remove --name '#{profile}'") }
  distros_api.each { |distro| $server.run("cobbler distro remove --name '#{distro}'") }
end
```

ByteOtter commented:
> There is more code that should be moved to the new Cobbler file like e.g.
> 
> ```ruby
> Given(/^distro "([^"]*)" exists$/) do |distro|
>   ct = CobblerTest.new
>   raise 'distro ' + distro + ' does not exist' unless ct.distro_exists(distro)
> end
> 
> Given(/^profile "([^"]*)" exists$/) do |profile|
>   ct = CobblerTest.new
>   raise 'profile ' + profile + ' does not exist' unless ct.profile_exists(profile)
> end
> 
> When(/^I remove kickstart profiles and distros$/) do
>   host = $server.full_hostname
>   # -------------------------------
>   # Cleanup kickstart distros and their profiles, if any.
> 
>   # Get all distributions: created from UI or from API.
>   distros = $server.run('cobbler distro list')[0].split
> 
>   # The name of distros created in the UI has the form: distro_label + suffix
>   user_details = $api_test.user.get_details('testing')
>   suffix = ":#{user_details['org_id']}:#{user_details['org_name'].delete(' ')}"
> 
>   distros_ui = distros.select { |distro| distro.end_with? suffix }.map { |distro| distro.split(':')[0] }
>   distros_api = distros.reject { |distro| distro.end_with? suffix }
>   distros_ui.each { |distro| $api_test.kickstart.tree.delete_tree_and_profiles(distro) }
>   # -------------------------------
>   # Remove profiles and distros created with the API.
> 
>   # We have already deleted the profiles from the UI; delete all the remaning ones.
>   profiles = $server.run('cobbler profile list')[0].split
>   profiles.each { |profile| $server.run("cobbler profile remove --name '#{profile}'") }
>   distros_api.each { |distro| $server.run("cobbler distro remove --name '#{distro}'") }
> end
> ```
> 
> Furthermore there are other steps in `command_steps.rb` which can be moved, too, e.g.:
> 
> ```ruby
> When(/^I restart cobbler on the server$/) do
>   $server.run('systemctl restart cobblerd.service')
> end
> 
> Then(/^the cobbler report should contain "([^"]*)" for "([^"]*)"$/) do |text, host|
>   node = get_target(host)
>   output, _code = $server.run("cobbler system report --name #{node.full_hostname}:1", check_errors: false)
>   raise "Not found:\n#{output}" unless output.include?(text)
> end
> 
> Then(/^the cobbler report should contain "([^"]*)" for cobbler system name "([^"]*)"$/) do |text, name|
>   output, _code = $server.run("cobbler system report --name #{name}", check_errors: false)
>   raise "Not found:\n#{output}" unless output.include?(text)
> end
> ```

found them. should be fixed now I hope.

ByteOtter commented on testsuite/features/step_definitions/cobbler_steps.rb:
```diff
@@ -0,0 +1,121 @@
+  raise "error while verifying isolinux.cfg parameter for Cobbler buildiso.\nLogs:\n#{result}" if code.nonzero?
+end
+
+When(/^I cleanup after Cobbler buildiso$/) do
```
should be fixed now

elariekerboull reviewed (approved):
I did not check if there were other Cobbler step in other files, otherwise lgtm

nodeg reviewed (approved):
LGTM. Thank you.

nodeg commented:
```ruby
When(/^I copy autoinstall mocked files on server$/) do
  target_dirs = "/var/autoinstall/Fedora_12_i386/images/pxeboot /var/autoinstall/SLES15-SP4-x86_64/DVD1/boot/x86_64/loader /var/autoinstall/mock"
  $server.run("mkdir -p #{target_dirs}")
  base_dir = File.dirname(__FILE__) + "/../upload_files/autoinstall/cobbler/"
  source_dir = "/var/autoinstall/"
  return_codes = []
  return_codes << file_inject($server, base_dir + 'fedora12/vmlinuz', source_dir + 'Fedora_12_i386/images/pxeboot/vmlinuz')
  return_codes << file_inject($server, base_dir + 'fedora12/initrd.img', source_dir + 'Fedora_12_i386/images/pxeboot/initrd.img')
  return_codes << file_inject($server, base_dir + 'mock/empty.xml', source_dir + 'mock/empty.xml')
  return_codes << file_inject($server, base_dir + 'sles15sp4/initrd', source_dir + 'SLES15-SP4-x86_64/DVD1/boot/x86_64/loader/initrd')
  return_codes << file_inject($server, base_dir + 'sles15sp4/linux', source_dir + 'SLES15-SP4-x86_64/DVD1/boot/x86_64/loader/linux')
  raise 'File injection failed' unless return_codes.all?(&:zero?)
end
```
Should also be part of the new Cobbler file.

srbarrios commented on testsuite/.rubocop_todo.yml:
```diff
@@ -394,39 +392,24 @@ Style/DocumentationMethod:
-    - 'features/support/namespaces/system.rb'
-    - 'features/support/twopence_init.rb'
-    - 'features/support/xmlrpc_client.rb'
+  Enabled: false
```
Yup. Right, let's allow it everywhere.

srbarrios commented on testsuite/.rubocop_todo.yml:
```diff
@@ -48,6 +48,7 @@ Layout/ClosingHeredocIndentation:
 Layout/EmptyLineAfterGuardClause:
   Exclude:
     - 'features/step_definitions/api_common.rb'
+    - 'features/step_definitions/cobbler_steps.rb'
```
Let's try to refactor the code causing this warning, so we don't need to include one more excluded entry.

srbarrios commented on testsuite/.rubocop_todo.yml:
```diff
@@ -358,15 +360,11 @@ Style/DateTime:
 # Cop supports --auto-correct.
 Style/DisableCopsWithinSourceCodeDirective:
   Exclude:
-    - 'features/step_definitions/command_steps.rb'
     - 'features/step_definitions/common_steps.rb'
-    - 'features/step_definitions/file_management_steps.rb'
-    - 'features/support/commonlib.rb'
-    - 'features/support/twopence_init.rb'
```
https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/DisableCopsWithinSourceCodeDirective
I would maybe set it as `Enable: false`.

But it would be good to vote about it :)

srbarrios commented on testsuite/.rubocop_todo.yml:
```diff
@@ -394,39 +392,24 @@ Style/DocumentationMethod:
 
 # Offense count: 16
 # Cop supports --auto-correct.
 Style/IfUnlessModifier:
```
We have this clause plenty of exclusions, maybe it's also time to just set it as `Enabled: false`

srbarrios commented on testsuite/.rubocop_todo.yml:
```diff
@@ -588,14 +571,15 @@ Style/StringHashKeys:
   Exclude:
     - 'Gemfile'
     - 'ext-tools/maintenance_crawler.rb'
+    - 'features/step_definitions/cobbler_steps.rb'
```
Let's try to clean it up to don't include this, and in the future we will slowly remove the rest of here too. That one about using single quotes I think it's quite important for Ruby style.

srbarrios commented on testsuite/features/step_definitions/cobbler_steps.rb:
```diff
@@ -0,0 +1,224 @@
+
+# cobbler daemon
+Given(/^cobblerd is running$/) do
+  ct = CobblerTest.new
```
As this is just a refactor moving stuff, all good.

But my two cents for the future, this ct = CobblerTest.new in several steps could cause problems when we run tests in parallel, ideally we should have a CobblerAPIHandler global instance, and instantiate it only once. Outside the steps maybe.
But again, out of scope of this PR.

srbarrios reviewed (approved):
LGTM.
Only thing I would do is try to fix those remarks from rubocop, but aside that the core of the PR looks good, so it can even be merged as it is.

nodeg commented:
I am general in favor of enabling more cops over time, fix our code and get rid of the RuboCop todo file and only define the things we do not like/do not want to use in the normal configuration file for RuboCop.
